### ISSUE TITLE
Boolean rule without default condition

### ISSUE THREAD

**Issue by anderseknert** [StyraInc/regal#152]
It would be a good addition to have a rule that warns on use of boolean rules that are missing a default condition. Obviously this isn't wrong per se, so this falls into the style category (I think). It should also be possible to provide a list of rules for which does not apply.

**Comment by srenatus** [comment]
Hm I'm on the fence here. I'd tend to lean into undefined; and many rules are boolean, basically all the simple ones...

```rego
default allow := false
allow if user_is_admin
allow if endpoint_is_public
allow if {
  # more complex stuff
}

user_is_admin if ...

endpoint_is_public if ...
```

Would you think that `user_is_admin` and `endpoint_is_public` needs a default rule?

**Comment by anderseknert** [comment]
I think you're right — this is really only valuable for "main" rules. The only way we have of knowing which ones are would be via the `entrypoint` annotation. We currently provide no recommendation for that, but perhaps we should (in a different rule that is)? Having one serves as good documentation, and it also makes it easier to compile the policy to different targets. If we made it a best practice to always have at least one, we could have this rule enforce that at least boolean _entrypoints_ have a default rule.

We'd need a new rule type for this I guess, which would check the aggregate of all files rather than each file individually.

**Comment by srenatus** [comment]
I guess there's another heuristic: boolean rules that aren't called anywhere. It would cover my example nicely; and still doesn't require entrypoint metadata.

**Comment by anderseknert** [comment]
Yeah I suppose so, but it could just be that they're unused and actually should be removed? 🤔

**Comment by srenatus** [comment]
hmm warn about rules that aren't declared entrypoint and aren't called...?

```rego
package foo
baz := true
```

```rego
package bar
q := data.foo[input]
```

I guess we're too dynamic to reliably do this... 🤔

**Comment by anderseknert** [comment]
Let's make entrypoint annotations a recommendation then. Best practices born out of laziness 😅

**Comment by anderseknert** [comment]
Thinking about it, this would make sense as a `custom` rule where the user could simply say "for any rule named `allow`, there must be an accompanying default rule". Same for entrypoint annotations: "for all modules, there must be at least one entrypoint defined".

See https://github.com/StyraInc/regal/issues/282 for some design ideas.
